net/wgpeerselector/files/usr/bin/wgpeerselector

```diff
@@ -0,0 +1,505 @@
+			table.insert(self.peers, WGPeer:new{
+				name = peer['.name'],
+				iface = self.iface,
+				public_key = peer.public_key,
+				endpoint = peer.endpoint,
+				allowed_ips = peer.allowed_ips or {},
+				persistent_keepalive = peer.persistent_keepalive or 0,
+			})
+		end
+	end)
+
+	if #self.peers < 1 then
+		log(syslog.LOG_ERR, "No peers defined. Exiting!", true)
+		os.exit(1)
+	end
+
+	log(syslog.LOG_INFO, 'Loaded '..tostring(#self.peers)..' peers.')
+end
+
+function WGPeerSelector:queue_all_peers()
+	-- copy table
+	for _, peer in pairs(self.peers) do
+		table.insert(self.queued_peers, peer)
+	end
+end
+
+function WGPeerSelector:queue_pop_random_peer()
+	return table.remove(self.queued_peers, math.random(#self.queued_peers))
+end
+
+function WGPeerSelector:try_connect_to_peer(peer, timeout)
+	if not peer:install_to_kernel() then
+		return false
+	end
+
+	sleep(timeout)
+	peer:update_stats_from_kernel()
+
+	local connection_successful = peer:has_recent_handshake()
+
+	if not connection_successful then
+		peer:uninstall_from_kernel()
+	end
+
+	return connection_successful
+end
+
+function WGPeerSelector:cleanup()
+	for _, peer in pairs(self.peers) do
+		-- if peer is installed
+		if peer:update_stats_from_kernel() then
+			peer:uninstall_from_kernel()
+		end
+	end
+end
+
+function WGPeerSelector:wait_for_iface()
+	local function check()
+		return sys_stat.stat('/sys/class/net/' .. self.iface)
+	end
+
+	if check() then
+		return false -- no waiting was needed
+	else
+		log(syslog.LOG_WARNING, 'Iface ' .. self.iface .. ' not present. Waiting for iface.')
+	end
+
+	while not check() do
+		sleep(3)
+	end
+
+	log(syslog.LOG_INFO, 'Iface ' .. self.iface .. ' found.')
+	return true -- waiting was needed
+end
+
+function WGPeerSelector:sync_ntp()
+	-- While we rely on system.ntp.servers, we do not rely on /etc/sysntpd,
+	-- because this daemon might have other firewall rules than sysntpd.
+	-- E.g. sysntpd might be configured to sync time only via this vpn.
+	local ntp_servers = uci:get("system", "ntp", "server")
+
+	-- As ntpd has an ugly retry behaviour if dns requests fail, we resolve
+	-- dns here by ourselves. Otherwise ntpd would block our program from
+	-- continuing.
+	local dns_results = {}
+	for i = 1, #ntp_servers do
+		local res, _, errcode = sys_sock.getaddrinfo(ntp_servers[i], nil, { socktype = sys_sock.SOCK_DGRAM })
+
+		if errcode then
+			log(syslog.LOG_WARNING, "DNS resolution of ntp server " .. ntp_servers[i]
+			    .. " failed: error code " .. tostring(errcode))
+		else
+			for j = 1, #res do
+				table.insert(dns_results, res[j].addr)
+			end
+		end
+	end
+
+	if #dns_results < 1 then
+		log(syslog.LOG_WARNING, "no (valid) dns result for ntp servers. skipping time sync.")
+		return false
+	end
+
+	local command = "ntpd -n -N -q -p '" .. table.concat(dns_results, "' -p '") .. "'"
+	local time_synced = os.execute(command) == 0
+
+	if not time_synced then
+		log(syslog.LOG_WARNING, "synchronizing time failed.")
```

Comment:
Timesync failed

Reply:
I don't see a change here, why is it marked as resolved?

Reply:
Because I implemented things locally and thought, I'll push this in a few minutes, then things occurred...